wasi-threads.wit.md

```diff
@@ -0,0 +1,40 @@
+# WASI threads API
+
+WASI threads is an API for thread creation.
+
+Its goal is to provide functions that allow implementation of a subset of `pthreads` API, but it doesn't aim to be 100% compatible with POSIX threads standard.
+
+
+## thread-id
+
+```wit
+/// Unique thread identifier.
+type thread-id = u32
+```
+
+## start-arg
+
+```wit
+/// A reference to data passed to the start function (`wasi_thread_start()`) called by the newly spawned thread.
+type start-arg = u32
+```
+
+## errno
+
+```wit
+/// Error codes returned by the `thread-spawn` function.
+enum errno {
+    /// TBD
+    eagain,
+}
+```
+
+## thread_spawn
+
+```wit
+/// Creates a new thread.
+thread-spawn: func(
+    /// A value being passed to a start function (`wasi_thread_start()`).
+    start-arg: start-arg,
+) -> result<thread-id, errno>
```

Comment:
i played a bit with wit-bindgen.

with this wit definition, do you mean to switch to the following core wasm abi?
```
// Unique thread identifier.
typedef uint32_t wasi_thread_id_t;

// A reference to data passed to the start function (`wasi_thread_start()`) called by the newly spawned thread.
typedef uint32_t wasi_start_arg_t;

// Error codes returned by the `thread-spawn` function.
typedef uint8_t wasi_errno_t;

#define WASI_ERRNO_EAGAIN 0

__attribute__((import_module("wasi"), import_name("thread-spawn")))
void __wasm_import_wasi_thread_spawn(int32_t, int32_t);

bool wasi_thread_spawn(wasi_start_arg_t start_arg, wasi_thread_id_t *ret, wasi_errno_t *err) {
  __attribute__((aligned(4)))
  uint8_t ret_area[8];
  int32_t ptr = (int32_t) &ret_area;
  __wasm_import_wasi_thread_spawn((int32_t) (start_arg), ptr);
  wasi_result_thread_id_errno_t result;
  switch ((int32_t) (*((uint8_t*) (ptr + 0)))) {
    case 0: {
      result.is_err = false;
      result.val.ok = (uint32_t) (*((int32_t*) (ptr + 4)));
      break;
    }
    case 1: {
      result.is_err = true;
      result.val.err = (int32_t) (*((uint8_t*) (ptr + 4)));
      break;
    }
  }
  if (!result.is_err) {
    *ret = result.val.ok;
    return 1;
  } else {
    *err = result.val.err;
    return 0;
  }
}
```

Reply:
i prefer to stick with a simple i32->i32 type at least for now.

Reply:
> i prefer to stick with a simple i32->i32 type at least for now.

i mean something like this:
```
/// Unique thread identifier or negative WASI errno.
type thread-id-or-negative-wasi-errno = s32

/// Creates a new thread.
thread-spawn: func(
    /// A value being passed to a start function (`wasi_thread_start()`).
    start-arg: start-arg,
) -> thread-id-or-negative-wasi-errno
```

btw, do you intend to use `thread-spawn` (with hyphen) for core wasm import?

Reply:
Ah, good catch. I think we do just mean s32 here (at least that is how I implemented things in Wasmtime). I don't know if we really need the new type but a doc comment indicating that negative return values indicate an error would be helpful.

As for what wit-bindgen version to use, it seems like we should target the latest published version available as a CI action.

Reply:
> Yes, I was aware of this change, that's why I highlighted it in the description. I know that wasi-libc, WAMR, and (I guess) wasmtime implement this method with `thread-id-or-negative-wasi-errno`. But I proposed using `result` instead as this is exactly what we want to achieve, and using negative value to indicate error looks like a workaround.
> 
> I'm ok with switching it back to `i32` here, but I'd rather prefer going with a long-term solution (which I believe is using `result`) and update all the places that already implement this function.

is this part of component-model already stable enough to start relying on the core wasm mapping?

anyway, as i stated earlier, i prefer to have the first version of wit definition here to be close to the current `README.md` and implementations as far as possible. and then start making changes on it. (like using `result`)
that way it would be far simpler to review for those of us not too familiar with component-model.

Reply:
> is this part of component-model already stable enough to start relying on the core wasm mapping?

I'd assume so given this is what's used in the preview2 prototyping; @sunfishcode , can you confirm that? If it's not, I agree we should switch back to `i32`.

> anyway, as i stated earlier, i prefer to have the first version of wit definition here to be close to the current README.md and implementations as far as possible. and then start making changes on it. (like using result)
that way it would be far simpler to review for those of us not too familiar with component-model.

I don't think this complicates the API significantly, so if we can get it right in the first place, why not doing that? We can ofc update README file accordingly later on as well.

Reply:
@loganek
is this the right understanding of the new abi you are proposing here?
https://github.com/WebAssembly/wasi-libc/pull/385

Reply:
Yes, that seems to match the generated output in https://bytecodealliance.github.io/wit-bindgen/ when passing the definition:
```
interface host-funcs {
  type start-arg = u32
  type thread-id = u32
  enum errno {
    /// TBD
    eagain,
  }
  thread-spawn: func(
    /// A value being passed to a start function (`wasi_thread_start()`).
    start-arg: start-arg,
  ) -> result<thread-id, errno>
}

world demo {
  import host: host-funcs 
}

```

The only question though is how stable the ABI is, hopefully I'll have answer for that soon.

Reply:
Yep, I don't want to keep this PR open forever so let's have a version that's compatible with current proposal; I'll open an issue and a PR for the follow-up discussion.

Reply:
Create a follow-up issue: https://github.com/WebAssembly/wasi-threads/issues/29